### PR TITLE
Session enhancement

### PR DESCRIPTION
**Pull request by DieMyst**
#204

### CODE REVIEW

File: js-client/src/Session.ts [hunk 1]
```diff
@@ -38,45 +43,103 @@ export class Session {
...
+    }
+
+    /**
+     * Closes session, cancel and delete all promises.
+     */
+    private closeSession(reason: string) {
+        if (!this.closed) {
+            this.closed = true;
+            this.closedStatus = reason;
+            this.cancelAllPromises(reason)
```
[the remark this answers is not in the document]
Comment (DieMyst) [reply js-client/src/Session.ts]: it closes after "random failed function"

File: js-client/src/main.ts [hunk 2]
```diff
@@ -18,55 +18,130 @@ import {TendermintClient} from "./TendermintClient";
...
-    let signer = new Signer(signingKey);
+        let result = await pr;
+        console.log("RESULT = " + JSON.stringify(result));
 
-    // `client002` is a default client for now
-    let client = new Client("client002", signer);
+    }
 
-    let s = engine.genSession(client);
+    async testCloseSession(count: number = 100) {
```
[the remark this answers is not in the document]
Comment (DieMyst) [reply js-client/src/main.ts]: it closes after "random failed function"

File: js-client/src/Session.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (DieMyst) [reply js-client/src/Session.ts]: only if the client will save Promise for future work. Is it incorrect?

File: js-client/src/Session.ts [hunk 3]
```diff
@@ -38,45 +43,103 @@ export class Session {
...
 
-        //todo there should be a manager that syncs calls and increments counter only after a successful request
-        this.counter = this.counter + 1;
+        let targetKey = this.targetKey(currentCounter);
 
-        return new ResultAwait(this.tm, targetKey).result()
+        // starts checking if a response has appeared
+        let resultAwait = new ResultAwait(this.tm, targetKey, this.sessionSummaryKey);
+        this.resultPromises = this.resultPromises.add(resultAwait);
+        let pr: Promise<Result> = resultAwait.result(this.config.requestsPerSec,
```
[the remark this answers is not in the document]
Comment (DieMyst) [reply js-client/src/Session.ts]: You are right that we need to store only last promise.
But I don't know anything about explicit `result()`. Why it is better to invoke it explicitly?

File: js-client/src/Session.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (DmiSergeev) [reply js-client/src/Session.ts]: If we forbid retrieving a Promise result after session closing, then:
* It's unclear, how the client would ensure that session closing (`@closeSession()` tx) was successful.
* Flexibility is limited: the client needs to sync every session to obtain its results, despite they are still stored and might be retrieved a bit earlier when they are really needed.

File: js-client/src/main.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (DmiSergeev) [reply js-client/src/main.ts]: I think it would be better to name it more precisely.

File: js-client/src/ResultAwait.ts [hunk 4]
```diff
@@ -14,60 +14,164 @@
...
+
+            // if result exists, return it
+            if (optionResult.nonEmpty) {
+                return optionResult.get
             }
 
-            await this.sleep(1000 / requestsPerSec)
+            // here the session is checked after the result is checked, as it may happen that the result is given,
+            // and after that the session was immediately closed
+            if (sessionInfo.nonEmpty) {
```
Comment (DmiSergeev) [inline js-client/src/ResultAwait.ts]: Could `Option.exists()` be used here?

File: js-client/src/ResultAwait.ts [hunk 5]
```diff
@@ -14,60 +14,164 @@
...
-                } else if (resultJs.Empty !== undefined) {
-                    return empty;
-                } else {
-                    return value(resultJs.Computed.value);
-                }
+            return pr;
+        } else {
+            return this.invokeResult
+        }
+
```
Comment (DmiSergeev) [inline js-client/src/ResultAwait.ts]: Triple empty line

File: js-client/src/Session.ts [hunk 6]
```diff
@@ -85,10 +127,25 @@ export class Session {
...
+    async sync(): Promise<Result> {
+        return this.lastResult.result();
+    }
+
+    /**
+     * Closes session locally and send a command to close the session on the cluster.
+     */
+    async close(reason: string = "") {
+        this.closeSession(reason);
+        return this.invoke("@closeSession")
```
Comment (DmiSergeev) [inline js-client/src/Session.ts]: The previous would set `closed` to `true`, so `invokeRaw` would throw. Maybe we need a test for this method.

Comment (DmiSergeev) [reply js-client/src/Session.ts]: Also, could we change `Session.closeSession` method's name somehow? This name doesn't inform what the difference with `Session.close`.

File: js-client/src/ResultAwait.ts (hunk 4 above)
Comment (DmiSergeev) [reply js-client/src/ResultAwait.ts]: I found it here https://www.npmjs.com/package/ts-option#optionaexistsp-_-a--boolean-boolean – isn't it a library you use?

File: js-client/src/Session.ts [hunk 7]
```diff
@@ -32,6 +32,7 @@ export class Session {
     private readonly config: SessionConfig;
     private counter: number;
     private lastResult: ResultAwait;
+    private closing: boolean;
```
Comment (DmiSergeev) [inline js-client/src/Session.ts]: We can probably replace these 2 boolean fields with a single field that is in 1 of 3 states.

**Pull request merged**